crates/re_query2/src/promise.rs

```diff
@@ -0,0 +1,130 @@
+use std::sync::Arc;
+
+use re_log_types::DataCell;
+
+// ---
+
+/// Uniquely identifies a [`Promise`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct PromiseId(pub(crate) re_tuid::Tuid);
+
+impl std::fmt::Display for PromiseId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl PromiseId {
+    /// Create a new unique [`PromiseId`] based on the current time.
+    #[allow(clippy::new_without_default)]
+    #[inline]
+    pub fn new() -> Self {
+        Self(re_tuid::Tuid::new())
+    }
+}
+
+// ---
+
+/// A [`Promise`] turns a source [`DataCell`] into a new [`DataCell`] with the helper of a
+/// [`PromiseResolver`].
+///
+/// Each promise is uniquely identified via a [`PromiseId`].
+///
+/// [`Promise`]s can be cloned cheaply.
+#[derive(Debug, Clone)]
+pub struct Promise {
+    id: PromiseId,
+    source: DataCell,
+}
```

Comment:
I wonder if Rust can do a niche optimization on this (so that `sizeof Option<Promise> == sizeof Promise`) 🤔 

`DataCell` is an `Arc`, so maybe if that is put first? 🤷

Reply:
`static_assertions::assert_eq_size!(Promise, Option<Promise>);` actually compile as-is :upside_down_face: